### ISSUE TITLE
MS Surface Detection Issues (Invoke-CMApplyDriverPackage.log)

### ISSUE THREAD

**Issue by TonyDewhurst1** [MSEndpointMgr/ModernDriverManagement#33]
Surface Laptop 3 is applying the wrong drivers (Surface Laptop 2) and it has missing drivers post build.

![image](https://user-images.githubusercontent.com/71382477/93328805-21e65900-f814-11ea-8892-a4ae4e18ea13.png)

**Comment by TonyDewhurst1** [comment]
[ApplyDriverPackage.log](https://github.com/MSEndpointMgr/ModernDriverManagement/files/5231837/ApplyDriverPackage.log)

**Comment by virtualtech516** [comment]
I am having the same problem with Surface Pro 7's. 
[ApplyDriverPackage.log](https://github.com/MSEndpointMgr/ModernDriverManagement/files/5385836/ApplyDriverPackage.log)

**Comment by Ar5had07** [comment]
Same problem.  It's detecting multiple models for 1 model.

**Comment by TonyDewhurst1** [comment]
@NickolajA Is there any update to this ticket 👍

**Comment by Ar5had07** [comment]
I am using the old script (webservice) for all Microsoft Surfaces.  Which is a shame as the old script cant be used over CMG (using the Admin Service)

**Comment by TonyDewhurst1** [comment]
Yes im in a similar position and want to use a CMG so atm im using SCCM native driver methods for surface devices and the rest this solution. @Ar5had07

**Comment by vartaxe** [comment]
Maybe its related to the sku list

https://docs.microsoft.com/en-us/surface/surface-system-sku-reference

And the missing sku

https://github.com/MicrosoftDocs/Devices-docs/issues/33

Maybe @NickolajA and @maurice-daly can give some inout here

**Comment by nodiaque** [comment]
Have the same problem. Script says it detected 6 packages, which should not be possible. When looking at the script, for a surface pro 4, it says the systekmsku used is surface_pro_4 and that it's using systemsku method. Surface pro 4 package have surface_pro_4 sku, but it also take other package like surface pro 5 and 6 which have surface_pro_1807 and stuff like that. Seems there's a logic error.

**Comment by vartaxe** [comment]
![image](https://user-images.githubusercontent.com/11176349/103884437-e3d0b600-50de-11eb-9bff-d155a190f246.png)
 
yes and sadly this will fail with error code 1 incorrect function 

`The task sequence execution engine failed executing the action (Dynamically Apply Drivers) in the group (Modern Driver Management) with the error code 1
Action output: ... icrosoftBIOSUpdate.ps1 range 0-5120
Downloaded file from http://my.org.com:443/NOCERT_SMS_DP_SMSPKG$/TIC00E40/sccm?/Invoke-MicrosoftBIOSUpdate.ps1 to C:\_SMSTaskSequence\Packages\TIC00E40\Invoke-MicrosoftBIOSUpdate.ps1 
VerifyContentHash: Hash algorithm is 32780
Content successfully downloaded at C:\_SMSTaskSequence\Packages\TIC00E40.
Resolved source to 'C:\_SMSTaskSequence\Packages\TIC00E40'
Working dir 'C:\_SMSTaskSequence\Packages\TIC00E40'
Command line for extension .exe is "%1" %*
Set command line: Run PowerShell Script
PowerShell command line is NOT shown in the log ('OSDLogPowerShellParameters' is NOT set to 'True')
Executing command line: Run PowerShell Script with options (0, 4)
Process completed with exit code 1
PowerShell command line returned code 1
ReleaseSource() for C:\_SMSTaskSequence\Packages\TIC00E40.
reference count 1 for the source C:\_SMSTaskSequence\Packages\TIC00E40 before releasing
Released the resolved source C:\_SMSTaskSequence\Packages\TIC00E40. The operating system reported error 1: Incorrect function. `

**Comment by vartaxe** [comment]
maybe can @merlinfrombelgium also help?

**Comment by Ar5had07** [comment]
We are all having the same problem. The authors have yet replied to it.

**Comment by TonyDewhurst** [comment]
I just kept to the conventional approach for MS devices and used the tool for Dell/Lenovo. Frustrating that we not got any comments on the issue.

**Comment by maurice-daly** [comment]
Hi Everyone,

I just wanted to come back to you on this issue. I am looking into this and aim to have it resolved next week. The solution is provided as open source and free of charge, and we have put a lot of effort into maintaining it, however we are also very busy with work and also need to make time for our families. So if we don't respond, its not that we don't care, your feedback helps drive the whole solution.

Maurice

**Comment by TonyDewhurst** [comment]
Thanks @maurice-daly  (;

**Comment by maurice-daly** [comment]
Could someone here also do me a favor and take screenshots of the package description which it is matching against. I've just been checking the logic here on a newly created Surface Laptop 2 package, and it is not being matched for the Surface Laptop 3. 

Also which version of the scripts are you using?

**Comment by merlinfrombelgium** [comment]
@maurice-daly I could help out with most surface laptop models, some older Pro models and the Book 3

**Comment by TonyDewhurst** [comment]
4.06 is the version I am currently using

**Comment by TonyDewhurst** [comment]
![image](https://user-images.githubusercontent.com/7440014/104351943-f5aac280-54fd-11eb-9d06-945fb20ec9a2.png)

**Comment by TonyDewhurst** [comment]
Just to give a bit of an update I cleaned out the previously installed drivers some months ago to go back to the native method for MS devices. I have reimported them this week to help with testing it looks like it is no longer an issue.

Could anyone else confirm by removing and re adding the driver sets to confirm this is the case.

My Scripts and versions have not changed since reported issue.

**Comment by TonyDewhurst** [comment]
Currently on Script 4.0.7 all i did was reacquire the driver packs and redownload them using the tool and it seems to be working.

I am currently downloading a few more just to see if it was a conflicting driver pack with a similar SKU, but I thought id let someone else check if its the same for them too.

**Comment by Ar5had07** [comment]
> Currently on Script 4.0.7 all i did was reacquire the driver packs and redownload them using the tool and it seems to be working.
> 
> 
> 
> I am currently downloading a few more just to see if it was a conflicting driver pack with a similar SKU, but I thought id let someone else check if its the same for them too. 
> 
> 
> 
> Just for clarity the script was exactly the same the only thing which has changed is i redownloaded the driver packs via the tool.

Could you show us your Logs please? I would like to see if it's being detected correctly. Thanks

**Comment by TonyDewhurst** [comment]
[ApplyDriverPackage.log](https://github.com/MSEndpointMgr/ModernDriverManagement/files/5851914/ApplyDriverPackage.log)
Sure see attached this is the surface laptop 3 which failed the first time.

**Comment by Ar5had07** [comment]
> [ApplyDriverPackage.log](https://github.com/MSEndpointMgr/ModernDriverManagement/files/5851914/ApplyDriverPackage.log)
> Sure see attached this is the surface laptop 3 which failed the first time.

That looks like it's working.  Can you post a screenshot of the packages' description please?

**Comment by TonyDewhurst** [comment]
![image](https://user-images.githubusercontent.com/7440014/105411635-8b93cb00-5c2b-11eb-8e02-ccf65ebcb2f7.png)

**Comment by TonyDewhurst** [comment]
actually it not correct if you look at the log 

**Computer SystemSKU determined as: Surface_Laptop_3_1867:1868**

Then look at the actual download its downloading the 



Attempting to download content files for matched driver package: Drivers - Microsoft Surface Laptop 3 - Intel 15 - Windows 10 1909 x64]LOG]!><time="11:24:24.2340" date="01-21-2021" component="ApplyDriverPackage" context="NT AUTHORITY\SYSTEM" type="1" thread="2532" file="">
<![LOG[ - Setting task sequence variable OSDDownloadDownloadPackages to: DCM00702]

**Comment by Ar5had07** [comment]
> actually it not correct if you look at the log
> 
> **Computer SystemSKU determined as: Surface_Laptop_3_1867:1868**
> 
> Then look at the actual download its downloading the Surface_Laptop_3_1872
> 
> ![image](https://user-images.githubusercontent.com/7440014/105412940-4ec8d380-5c2d-11eb-8da7-fbc0ae9d7db9.png)
> 
> Attempting to download content files for matched driver package: Drivers - Microsoft Surface Laptop 3 - Intel 15 - Windows 10 1909 x64]LOG]!><time="11:24:24.2340" date="01-21-2021" component="ApplyDriverPackage" context="NT AUTHORITY\SYSTEM" type="1" thread="2532" file="">
> <![LOG[ - Setting task sequence variable OSDDownloadDownloadPackages to: DCM00702]

Yeah for that model it's not working.  I just tested script version 4.0.9.  It's downloading the wrong Surface Laptop 3.

Here are my observations:
I am running in DEBUG mode to figure out what's going on.  Here is my log.  In red, it says it's using SystemSKU but in blue it's not using the SystemSKU to match the models.
![image](https://user-images.githubusercontent.com/46709367/105413499-18d81f00-5c2e-11eb-8016-720129ddbb99.png)

**Comment by TonyDewhurst** [comment]
actually its not working in my log its has detected sku

 **Computer SystemSKU determined as: Surface_Laptop_3_1867:1868]**

but its downloading



> > actually it not correct if you look at the log
> > **Computer SystemSKU determined as: Surface_Laptop_3_1867:1868**
> > Then look at the actual download its downloading the Surface_Laptop_3_1872
> > ![image](https://user-images.githubusercontent.com/7440014/105412940-4ec8d380-5c2d-11eb-8da7-fbc0ae9d7db9.png)
> > Attempting to download content files for matched driver package: Drivers - Microsoft Surface Laptop 3 - Intel 15 - Windows 10 1909 x64]LOG]!><time="11:24:24.2340" date="01-21-2021" component="ApplyDriverPackage" context="NT AUTHORITY\SYSTEM" type="1" thread="2532" file="">
> > <![LOG[ - Setting task sequence variable OSDDownloadDownloadPackages to: DCM00702]
> 
> Yeah for that model it's not working. I just tested script version 4.0.9. It's downloading the wrong Surface Laptop 3.
> 
> Here are my observations:
> I am running in DEBUG mode to figure out what's going on. Here is my log. In red, it says it's using SystemSKU but in blue it's not using the SystemSKU to match the models.
> ![image](https://user-images.githubusercontent.com/46709367/105413499-18d81f00-5c2e-11eb-8016-720129ddbb99.png)

Yes good spot

**Comment by Ar5had07** [comment]
I managed to fix the issue, it's dirty but it works and that's all I care about.  I will post what I did tomorrow.

**Comment by Ar5had07** [comment]
So far, the only model that's causing issues with detection (from my testing) is the Surface Laptop 3.  I have made this quick modification to the script and the packages's description in order to be detected correctly.

I added the following for Microsoft under Get-ComputerData

"*Microsoft*" {
				$ComputerDetails.Manufacturer = "Microsoft"
				$ComputerDetails.Model = (Get-WmiObject -Class "Win32_ComputerSystem" | Select-Object -ExpandProperty Model).Trim()
				$ComputerDetails.SystemSKU = Get-WmiObject -Namespace "root\wmi" -Class "MS_SystemInformation" | Select-Object -ExpandProperty SystemSKU | Foreach-object { switch ( $_ ) {
								"Surface_Laptop_3_1867:1868" { Return "SurfaceLaptop3Intel13" }
								"Surface_Laptop_3_1873" { Return "SurfaceLaptop3AMD13" }
								"Surface_Laptop_3_1872" { Return "SurfaceLaptop3Intel15" }
							       }
								}
				}

(sorry I tried inserting it as code, but it was a big mess)

And I renamed the Surface Laptop 3 packages to Surface Lap 3, because it's using the name as a detection method.  I also renamed the descriptions to SurfaceLaptop3Intel15, SurfaceLaptop3Intel13, SurfaceLaptop3Intel15.  

So during OSD, if you are building model Surface_Laptop_3_1867:1868, the script will return SurfaceLaptop3Intel13 and that will match the description of your packages, which will result in the correct package being downloaded.

I ran this

`.\Invoke-CMApplyDriverPackage-v4.0.9.ps1 -DebugMode -Endpoint "xxx.NET" -UserName "yyy" -Password "yyy" -TargetOSVersion 1909 -Manufacturer "Microsoft" -ComputerModel "Surface Laptop 3" -SystemSKU "SurfaceLaptop3Intel15"`

and this is the output

![image](https://user-images.githubusercontent.com/46709367/105491171-cdfdec00-5cad-11eb-9f28-a5ac272792f9.png)

It's messy but it works, and that's good enough with me.  Let me know if it works for you.

**Comment by tewebb** [comment]
I just wanted to loop back on this, I'm still seeing a multiple package found issue with the Surface Pro 7+ (also, the driver importer doesn't show 7+ drivers). I've attached the associate log file.
[ApplyDriverPackage.log](https://github.com/MSEndpointMgr/ModernDriverManagement/files/6504289/ApplyDriverPackage.log)

**Comment by 20H221H1** [comment]
Where do you guys find the log file ?

**Comment by Ar5had07** [comment]
> I just wanted to loop back on this, I'm still seeing a multiple package found issue with the Surface Pro 7+ (also, the driver importer doesn't show 7+ drivers). I've attached the associated log file.
> 
> [ApplyDriverPackage.log](https://github.com/MSEndpointMgr/ModernDriverManagement/files/6504289/ApplyDriverPackage.log)
> 
> 

I posted the solution above.

**Comment by Ar5had07** [comment]
> Where do you guys find the log file ?

C:/ProgramFiles/Msendpoingmgr/logs

**Comment by 20H221H1** [comment]
C:/ProgramFiles/Msendpoingmgr/logs - contains the "DriverAutomationTool.log" file

Where can I find the "ApplyDriverPackage.log" file ?

**Comment by Ar5had07** [comment]
> C:/ProgramFiles/Msendpoingmgr/logs - contains the "DriverAutomationTool.log" file
> 
> Where can I find the "ApplyDriverPackage.log" file ?

ApplyDriverPackage.log - during TS it's located in X:Temp\Logs\ApplyDriverPackage.log
During Full OS or DEBUD - C:\Windows\Temp\ApplyDriverPackage.log
